fastavro/util.py

```diff
@@ -0,0 +1,72 @@
+# cython: auto_cpdef=True
+'''Utility functions'''
+
+import sys
+from functools import wraps
+
+if sys.version_info < (3,):
+    INTEGER_TYPES = (int, long,)  # flake8: noqa
+else:
+    INTEGER_TYPES = (int,)
+
+
+def push_path(path, name):
+    if path is None:
+        return [name]
+    else:
+        path.append(name)
+        return path
+
+
+def pop_path(path):
+    if path is None:
+        return None
+    else:
+        path.pop()
+        return path
+
+
+def path_string(path):
+    def format_item(item):
```

Comment:
This will define a new function every time the function is called and potentially call it. In order maintain speed, I prefer we'll unroll this. Also I `[]` will be the empty path and not `None` it'll make things easier.

```python
fmt = '[%s]' if isinstance(item, INTEGER_TYPE) else '%s'
return '.'.join(fmt % item for item in path)
```

Also what's the motivation for `[1]` display? Maybe we can use `%r` to distinguish between `int` and `str`.